receiver/kafkareceiver/consumer_franz.go

```diff
@@ -0,0 +1,476 @@
+					zap.String(attrPartition, strconv.Itoa(int(partition))),
+				),
+				attrs: attribute.NewSet(
+					attribute.String(attrInstanceName, c.id.String()),
+					attribute.String(attrTopic, topic),
+					attribute.String(attrPartition, strconv.Itoa(int(partition))),
+				),
+			}
+		}
+	}
+}
+
+// lost must be set as a kgo.OnPartitionsLost and kgo.OnPartitionsReassigned
+// callbacks. Ensures that partitions that are lost (see kgo.OnPartitionsLost
+// for more details) or reassigned (see kgo.OnPartitionsReassigned for more
+// details) have their partition consumer stopped.
+// This callback must finish within the re-balance timeout.
+func (c *franzConsumer) lost(_ context.Context, _ *kgo.Client, lost map[string][]int32) {
+	c.mu.Lock()
+	defer c.mu.Unlock()
+	for topic, partitions := range lost {
+		for _, partition := range partitions {
+			tp := topicPartition{topic: topic, partition: partition}
+			c.telemetryBuilder.KafkaReceiverPartitionClose.Add(context.Background(),
+				1, metric.WithAttributes(attribute.String(attrInstanceName, c.id.Name())),
+			)
+			delete(c.assignments, tp)
+		}
+	}
+}
+
+// handleMessage is called on a per-partition basis.
+func (c *franzConsumer) handleMessage(ctx context.Context, msg kafkaMessage, pc *pc) error {
+	defer pc.resetBackoff()
+
+	for {
+		err := c.consumeMessage(ctx, msg, pc.attrs)
+		if err == nil {
+			return nil // Successfully processed.
+		}
+		// In the future, with Consumer Share Groups, messages not processed
+		// within a configurable timeout, are re-delivered to the consumer in
+		// the Share group, however, at the time of writing this feature isn't
+		// yet GA nor widely deployed.
+		// Share groups are more in line with observability and OTel collector
+		// use cases than traditional consumer groups.
+		// https://cwiki.apache.org/confluence/display/KAFKA/KIP-932%3A+Queues+for+Kafka.
+		if pc.backOff != nil && !consumererror.IsPermanent(err) {
+			backOffDelay := pc.getNextBackoff()
+			if backOffDelay != backoff.Stop {
+				pc.logger.Info("Backing off due to error from the next consumer.",
+					zap.Error(err),
+					zap.Duration("delay", backOffDelay),
+				)
+				select {
+				case <-ctx.Done():
+					return nil
+				case <-time.After(backOffDelay):
+					continue
+				}
+			}
+			pc.logger.Info("Stop error backoff because the configured max_elapsed_time is reached",
+				zap.Duration("max_elapsed_time", pc.backOff.MaxElapsedTime),
+			)
+		}
+		if c.config.MessageMarking.After && !c.config.MessageMarking.OnError {
+			// Only return an error if messages are marked after successful processing.
+			return err
+		}
+		pc.logger.Error("failed to consume message, skipping due to message_marking config",
+			zap.Error(err),
+			zap.Int64("offset", msg.offset()),
+		)
+		return nil
+	}
+}
+
+// pc represents the partition consumer shared information.
+type pc struct {
+	logger *zap.Logger
+	attrs  attribute.Set
+
+	backOff      *backoff.ExponentialBackOff
+	backOffMutex sync.Mutex
```

Comment:
Why does this need locking? Shouldn't there only ever be a single goroutine per topic-partition?

Reply:
ah not anymore. C/P issue.